[IssueTracker] Update searchQuery in snippet (like in github)

BillyCottrell opened the pull request:
So I made some changes to the snippet so that when you press "open", "closed" or any label it updates the search query to that corresponding value.
It also keeps track of the current label or status so you don't lose one or the other.
This way it should act like you are using github to quickly filter issues.

BillyCottrell committed:
Update searchQuery (like in github):
- when pressing status buttons
- when pressing a label

BillyCottrell committed:
Added Label search from issue
- When pressing a label from an issue it now opens up the issue tracker and searches for that label
- Also fixed pressing open closed button

BillyCottrell commented:
I fixed a bug where when you pressed the open button multiple times it switched between open or closed (same for closed button).
Also implemented the ability that when you have an issue and you click the label you now navigate to the issue tracker list and auto search open issues for that label.

ljavuras commented:
Great addition, I'll be testing these out!

ljavuras commented:
@BillyCottrell Sorry for the late reply, first off, thank you for the PR, it's really appreciated.

I've found some issues in your PR, if you're interested in fixing those, I'll merge after your fix, otherwise I'll merge directly.

Magisker commented:
@ljavuras ye sure no problem, just let me know what issues there are and ill fix it rn

BillyCottrell commented:
@ljavuras ye sure no problem, let me know what issues there are and ill fix it as soon as possible (ps sorry, i commented with wrong account before)

ljavuras commented:
@BillyCottrell tysm! I've started a review and added comments, please check!

Status buttons can fail when search query can have no status filters, i.e., `is:open` or `is:closed`, which will search for both open and closed issues.

The way label chips implemented will also fail when no status filters are present, and will always erase existing labels.

I've addressed both issues in my own active vault: [IssueTracker/view.js](https://gist.github.com/ljavuras/d47247a1399cf3b34a6ceb2e06796626).
It depends heavily on external scripts, so not very sharable, but I hope it can help you a little bit.

BillyCottrell commented:
@ljavuras so i have been trying to setup a clean separate environment for it but i can't get it to work for some reason (with just the changes i made) like I can't create an issue as it fails when executing app.vault.cachedread to read the files content.
So then I tried out the view.js file you shared but I can't get that one to work either as "Obsidian" which you load in from cjs is undefined.
So I am trying to fix it but ill prob have to spend some more time on it to just get it to work again.

ljavuras commented:
@BillyCottrell Feel free to reach out for any help, I've updated the [gist](https://gist.github.com/ljavuras/d47247a1399cf3b34a6ceb2e06796626) to include 2 CustomJS scripts `Obsidian.js` and `Templater.js`.

BillyCottrell commented:
@ljavuras thx, could you also add the updated issue/view.js file then, cause its also missing the customjs code you prob added so it can be imported.

PS: To be honest I think it might be better for a separate issuetracker version for customjs support, because it will still work without customjs (or atleast i got it too work again without customjs). That way you have a base version and a customjs supported version. (or perhaps do a check whether customjs is available or not)

ljavuras commented:
I'm not sure what you need because there's no single `view.js` for issues. You might have to change the issue template that issue tracker is using, tho. I have different template names in my private vault.

It's not that easy to share a "CustomJS version issue tracker", because all of my scripts are tightly integrated together.

For example, my project, issue tracker, and kanban works cohesively [like this](https://imgur.com/130Lxe0). And there is a [template updater](https://imgur.com/Lkh2Npn) entangles with all three of them because they all use template.

That's why I didn't share as much on this repo, it's impossible to manage two repos with the frequency I am developing in my active vault.

BillyCottrell commented:
oh okay, i just copied the files from the repo again which contains the view.js and view.css for both issuetracker as each issue individually. I asked since it was creating an error for being unable to import that file.

I removed those files but the issue remains except its for the view.js of the IssueTracker:
![image](https://github.com/user-attachments/assets/dee32b1a-a627-48c1-aeb1-2dae0f99dfcc)

ljavuras commented:
I'm seeing CustomJS complaining it couldn't import `view.js`, and that is how it should be! `view.js` are Dataview scripts and are meant to be called by`dv.view()`.

BillyCottrell commented:
well that's what i am doing, which should be fine:
![image](https://github.com/user-attachments/assets/b3aea859-956c-4652-9544-fa89594f67e8)

BillyCottrell commented:
So I changed customjs to only look at the files that are under the directory of customjs itself and now the console error is gone, but dataview is still giving that errormessage for some reason.
And I still can't create an issue either as I get these errors aswell:
![image](https://github.com/user-attachments/assets/f41fd7e8-9997-46de-99e2-7665a0f1137b)


Guess that will be for tomorrow then instead

ljavuras commented:
Not sure about the `read properties of null`, but the dataviewjs error can be safely ignored, and delete `dv.view("...path/to/Navigation")`. That part of the view relies on `Project.js` and `Kanban.js` script to render properply:
![image](https://github.com/user-attachments/assets/761e99b2-a945-4d24-9e06-3bc0e066776d)

BillyCottrell commented:
Okay so I have tried so solve the issues on this pr ignoring your customjs setup as i still can't figure it out.
And so far the first issue is resolved, but the second one about the label chips i don't really understand what is failing about it.

BillyCottrell committed:
Fixed status and label toggling

BillyCottrell commented:
Okay so I think most bugs related to status/label selection is now fixed. I didn't implement it the way you did since the current version uses a js file for the issue tracker and a js file for the issue itself aswell.

BillyCottrell committed:
forgot to remove console.logs from testing

ljavuras commented:
@BillyCottrell Thank you! I'll check asap.

ljavuras commented:
@BillyCottrell I've been trying to test your pull request, but I couldn't fetch your branch locally.
```console
❯ git fetch origin pull/5/head:main
From https://github.com/ljavuras/obsidian-power-tools
 ! [rejected]        refs/pull/5/head -> main  (non-fast-forward)
```
Fetch into branches other than main doesn't help either, I'll have to test a clone of your repo if I can't figure this out.

BillyCottrell commented:
@ljavuras that command does indeed not work because your main has diverged so git can't automatically overwrite it without potentially losing commits.
So I did this and it worked:
```
> git fetch origin pull/5/head:pr-5
remote: Enumerating objects: 30, done.
remote: Counting objects: 100% (30/30), done.
remote: Compressing objects: 100% (15/15), done.
remote: Total 24 (delta 16), reused 15 (delta 9), pack-reused 0 (from 0)
Unpacking objects: 100% (24/24), 11.58 KiB | 515.00 KiB/s, done.
From https://github.com/ljavuras/obsidian-power-tools
 * [new ref]         refs/pull/5/head -> pr-5
 
 > git checkout pr-5
 Switched to branch 'pr-5'
```

ljavuras commented:
@BillyCottrell That doesn't work for me either, both branches branched out from `main` and `develop` receive rejected message.
```console
❯ git fetch origin pull/5/head:pr-5
From https://github.com/ljavuras/obsidian-power-tools
 ! [rejected]        refs/pull/5/head -> pr-5  (non-fast-forward)
```

ljavuras commented:
@BillyCottrell I've tested your issue tracker and noticed some wonky behavior:
1. When query only searches label, e.g., `lable:feature`, clicking either status button will not modify the query
2. When search query has multiple label queries, clicking label will only replace the first label, e.g., `label:issue label:bug` -> clicks bug label -> `label:bug label:bug`
3. Clicking labels in issue doesn't bring me back to issue tracker, I couldn't get this feature to work.

I extracted the [diffs](https://gist.github.com/ljavuras/26ed33a66dd410846ab831499d816b6c) I made to implement some of these features, hope these can help.

That said, many previous mentioned problems are fixed, and it's a huge improvement compared to the version you forked, so I will merge if you don't intend to fix anymore.

BillyCottrell committed:
Update view.js

- Fixed statusBtn no longer working when only label filter is present.
- Fixed label filtering now removes all label filters when clicking on a label chip

BillyCottrell commented:
@ljavuras So I fixed the first two issues, but the third one works for me.
Just to be sure did you open the issue note itself or did you try to click the label from the preview?
I don't know how to fix the preview one to be honest since if you hover over the issue title and then click on a label chip it won't work.
Except if you click inside the preview and then click on the label it does set the search value in the issue tracker.
It just doesn't update ths issue list aslong as you are viewing the preview, but once you leave the preview then the issue tracker will update the issue list so it kinda works but its not intuitive, but no clue how to fix that one rn.

ljavuras commented:
@BillyCottrell I've tested and it works much smoother!

Clicking labels in issue didn't work because my README named the issue tracker note `Issue Tracker.md`, but both examples I provided don't follow this schema, so these lines didn't work:
```js
// Issue Tracker/Issue Tracker/Issue/view.js:L52-L55
        this.el.onclick = () => {
            // if the location is used as explained in the examples this should work
            var issueTrackerPath =
              app.workspace.getActiveFile().parent.parent.path + "/Issue Tracker.md";
```

As for the problem you described, perhaps it's because when clicking in hover preview, the active file is the issue tracker, not the issue, so these lines fails to get the correct issue tracker note:
```js
// Issue Tracker/Issue Tracker/Issue/view.js:L54-L56
      var issueTrackerPath =
        app.workspace.getActiveFile().parent.parent.path + "/Issue Tracker.md";
      var issueTrackerFile = getFile(issueTrackerPath);
```

BillyCottrell commented:
@ljavuras oh okay, glad it works!
The preview issue is like you said probably those lines that are failing so just a check whether the active file is the "Issue Tracker.md" file or not would solve it. But that's not that a big of an issue anyways, and is entirely up to you if you want me to fix this for you or just skip it so you can start the merge.

ljavuras commented:
@BillyCottrell Merged! Thank you.
